### PR TITLE
Sage Intacct Redesign : Export Settings

### PR DESCRIPTION
**Pull request by anishfyle**
Screenshot 2023-07-24 at 11 18 57 AM

### CODE REVIEW

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                    [label]="'Export Reimbursable Expenses'" 
+                    [subLabel]="'Enable this to export the reimbursable expenses from Fyle. If not enabled, any <b>out-of-pocket</b> expenses will not be exported to Sage Intacct .'" 
+                    [formControllerName]="'reimbursableExpense'"></app-configuration-toggle-field>
+                </div>
+                <div *ngIf="exportSettingsForm.value.reimbursableExpense">
+                    <div class="tw-mt-12-px tw-bg-white tw-border tw-border-solid tw-border-separator tw-rounded-12-px">
+                        <app-configuration-select-field
+                        [form]="exportSettingsForm"
+                        [isFieldMandatory]="true"
+                        [mandatoryErrorListName]="'export state'"
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:27]: export type

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                        [label]="'How should the expenses be exported?'"
+                        [subLabel]="'Choose the type of transaction record that you would like to create in Sage Intacct while exporting expenses from Fyle. Click here for more details.'"
+                        [options]="reimbursableExportTypes"
+                        [iconPath]="'expense'"
+                        [placeholder]="'Select expense export module'"
+                        [formControllerName]="'reimbursableExportType'"></app-configuration-select-field>
+                        <div *ngIf="exportSettingsForm.value.reimbursableExportType==='JOURNAL_ENTRY'">
+                            <div>
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:35]: this div can be deleted

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+                                <app-configuration-select-field
+                                [form]="exportSettingsForm"
+                                [isFieldMandatory]="true"
+                                [mandatoryErrorListName]="'export state'"
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:39]: update list name

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                                [subLabel]="'The integration will credit the account selected here for Reimbursable Expenses exported as Journal Entries.'"
+                                [options]="['']"
+                                [iconPath]="'copy-v2'"
+                                [placeholder]="'Select representation'"
+                                [formControllerName]="'reimbursableExportType'"></app-configuration-select-field>
+                            </div>
+                        </div>
+                    
+                    </div>
+                    <div *ngIf="exportSettingsForm.controls.reimbursableExportType.value?.value==='JOURNAL_ENTRY'" class="tw-mt-12-px tw-bg-white tw-border tw-border-solid tw-border-separator tw-rounded-12-px">
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:50]: exportSettingsForm.value.reimbursableExportType

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+      label: 'Bill',
+      value: ReimbursableExpensesObject.BILL
+    },
+    {
+      label: 'Journal Entry',
+      value: ReimbursableExpensesObject.JOURNAL_ENTRY
+    }
+  ];
+
+  EntityNameOptions: ExportSettingFormOption[] = [
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts:118]: employeeFieldOptions

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                        <app-configuration-select-field
+                        [form]="exportSettingsForm"
+                        [isFieldMandatory]="true"
+                        [mandatoryErrorListName]="'export state'"
+                        [label]="'How are your Employees represented in Sage Intacct?'"
+                        [subLabel]="'Select how you represent your employees in Sage Intacct. This would help to export the expenses to the correct record in Sage Intacct.'"
+                        [options]="EntityNameOptions"
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:57]: employeeFieldOptions

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+                        [iconPath]="'copy-v2'"
+                        [placeholder]="'Select representation'"
+                        [formControllerName]="'employeeFieldMapping'"></app-configuration-select-field>
+                    </div>
+                    <div *ngIf="exportSettingsForm.controls.employeeFieldMapping.value==='EMPLOYEE'" class="tw-mt-12-px tw-bg-white tw-border tw-border-solid tw-border-separator tw-rounded-12-px">
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:62]: don't involve form.control anywhere, change it to form.value.<>

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                        <app-configuration-select-field
+                        [form]="exportSettingsForm"
+                        [isFieldMandatory]="true"
+                        [mandatoryErrorListName]="'export state'"
+                        [label]="'How should Employees in Fyle be mapped to Employees in Sage Intacct?'"
+                        [subLabel]="'Automatically map the employees in Fyle to their corresponding records in Sage Intacct based on a unique parameter.'"
+                        [options]="[]"
+                        [iconPath]="'copy-v2'"
+                        [placeholder]="'Select representation'"
+                        [formControllerName]="'employeeFieldMapping'"></app-configuration-select-field>
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:72]: add a new controller for auto map

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+                        [form]="exportSettingsForm"
+                        [isFieldMandatory]="true"
+                        [mandatoryErrorListName]="'export state'"
+                        [label]="'How should Employees in Fyle mapped to Vendors in Sage Intacct?'"
+                        [subLabel]="'Automatically map the vendors in Fyle to their corresponding records in Sage Intacct based on unique parameter.'"
+                        [options]="[]"
+                        [iconPath]="'copy-v2'"
+                        [placeholder]="'Select representation'"
+                        [formControllerName]="'employeeFieldMapping'"></app-configuration-select-field>
+                    </div>
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:85]: not needed

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                        [iconPath]="'questionmark'"
+                        [placeholder]="'Select export state'"
+                        [formControllerName]="'reimbursableExpenseState'"></app-configuration-select-field>
+                    </div>
+                    <div class="tw-mt-12-px tw-bg-white tw-border tw-border-solid tw-border-separator tw-rounded-12-px">
+                        <app-configuration-select-field
+                        [form]="exportSettingsForm"
+                        [isFieldMandatory]="true"
+                        [mandatoryErrorListName]="'export state'"
+                        [label]="'Set the expense/bill date as?'"
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:103]: expense/bill should be dynamic, check QBO code for example

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+                        [placeholder]="'Select expense export module'"
+                        [formControllerName]="'cccExportType'"></app-configuration-select-field>
+                    </div>
+                    <div class="tw-mt-12-px tw-bg-white tw-border tw-border-solid tw-border-separator tw-rounded-12-px">
+                        <app-configuration-select-field
+                        [form]="exportSettingsForm"
+                        [isFieldMandatory]="true"
+                        [mandatoryErrorListName]="'export state'"
+                        [label]="'At which state should the expenses be ready to export from Fyle?'"
+                        [subLabel]="'You can export expenses either when they are awaiting payment after approval (Payment processing) or when the payment has been settled (Paid).'"
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:151]: Change Payment Processing to Processing and Paid to Closed

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+          value: ExpenseState.PAYMENT_PROCESSING
+        },
+        {
+          label: this.is_simplify_report_closure_enabled ? 'Closed' : 'Paid',
+          value: ExpenseState.PAID
+        }
+      ];
+    }
+
+    private setCreditCardExpenseGroupingDateOptions(cccExportType: CorporateCreditCardExpensesObject, cccExportGroup: ExpenseGroupedBy) : void {
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts:164]: rename this function better

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+      this.exportSettingService.getExportSettings().subscribe((exportSettingResponse : ExportSettingGet) => {
+        this.exportSettings = exportSettingResponse;
+        this.is_simplify_report_closure_enabled = this.exportSettings?.is_simplify_report_closure_enabled;
+        this.setUpExpenseStates();
+        this.setupCCCExpenseGroupingDateOptions();
+
+        this.exportSettingsForm = this.formBuilder.group({
+          reimbursableExportType: [this.exportSettings?.reimbursable_expenses_export_type],
+          reimbursableExpense: [this.exportSettings?.reimbursable_expenses_export_type ? true : false],
+          reimbursableExportGroup: [this.exportSettings?.reimbursable_expense_grouped_by ? this.exportSettings?.reimbursable_expense_grouped_by : null],
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts:209]: this needs to be changed similar to QBO

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+          reimbursableExportDate: [this.exportSettings?.reimbursable_expense_date ? this.exportSettings?.reimbursable_expense_date : null],
+          employeeFieldMapping: [this.exportSettings?.employeeFieldMapping ? this.exportSettings?.employeeFieldMapping : null],
+          creditCardExpense: [this.exportSettings?.credit_card_expense_export_type ? true : false],
+          cccExportType: [this.exportSettings?.credit_card_expense_export_type ? this.exportSettings?.credit_card_expense_export_type : null],
+          cccExportGroup: [this.exportSettings?.credit_card_expense_grouped_by ? this.exportSettings?.credit_card_expense_grouped_by : this.expenseGroupingFieldOptions[1].value],
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts:214]: this needs to be changed similar to QBO

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+          cccExportDate: [this.exportSettings?.credit_card_expense_date ? this.exportSettings?.credit_card_expense_date : this.cccExpenseGroupingDateOptions[0].value],
+          bankAccount: [this.exportSettings?.bank_account_name ? this.exportSettings?.bank_account_name : null],
+          cccEntityName: [this.exportSettings?.credit_card_entity_name_preference ? this.exportSettings?.credit_card_entity_name_preference : null],
+          cccAccountName: [this.exportSettings?.credit_card_account_name ? this.exportSettings?.credit_card_account_name : null],
+          reimbursableExpenseState: [this.exportSettings?.reimbursable_expense_state ? this.exportSettings?.reimbursable_expense_state : null],
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts:219]: we call it expense_state

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+          cccExpenseState: [this.exportSettings?.credit_card_expense_state ? this.exportSettings?.credit_card_expense_state : null]
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts:220]: ccc_expense_state

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html [hunk 17]
```diff
@@ -43,53 +42,39 @@
...
-                        [isFieldMandatory]="true"
-                        [mandatoryErrorListName]="'export state'"
-                        [label]="'How should Employees in Fyle mapped to Vendors in Sage Intacct?'"
-                        [subLabel]="'Automatically map the vendors in Fyle to their corresponding records in Sage Intacct based on unique parameter.'"
-                        [options]="[]"
-                        [iconPath]="'copy-v2'"
+                        [iconPath]="'threepeople'"
                         [placeholder]="'Select representation'"
-                        [formControllerName]="'employeeFieldMapping'"></app-configuration-select-field>
+                        [formControllerName]="'auto_map_employees'"></app-configuration-select-field>
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:69]: `autoMapEmployees` camel case

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 18]
```diff
@@ -137,6 +137,23 @@ export class ConfigurationExportSettingsComponent implements OnInit {
...
+        return exportGroup ? exportGroup : ExpenseGroupingFieldOption.CLAIM_NUMBER;
+      }
+  
+      return '';
+    }
+
+    getExportType(exportType: IntacctReimbursableExpensesObject | CorporateCreditCardExpensesObject): string {
+      const lowerCaseWord = exportType.toLowerCase();
+  
+      return lowerCaseWord.charAt(0).toUpperCase() + lowerCaseWord.slice(1);
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts]: can use angular pipes here directly, 1 example from QBO
Screenshot 2023-08-16 at 1 39 40 PM

**Review by ashwin1111 — APPROVED** [review]

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html (hunk 17 above)
Comment (anishfyle) [reply src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.html:69]: completed in 2nd PR 👍

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 19]
```diff
@@ -205,28 +206,6 @@ export class ConfigurationExportSettingsComponent implements OnInit {
       ];
     }
 
-    private setCCExpenseDateOptions(cccExportType: CorporateCreditCardExpensesObject, cccExportGroup: string) : void {
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts]: why is this removed?

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 20]
```diff
@@ -267,31 +247,37 @@ export class ConfigurationExportSettingsComponent implements OnInit {
           this.exportSettingsForm.controls.reimbursableExportGroup.setValue(null);
           this.exportSettingsForm.controls.reimbursableExportDate.setValue(null);
         }
-        this.reimbursableExportWatcher();
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts]: why not create watcher at the start?

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 21]
```diff
@@ -303,20 +289,17 @@ export class ConfigurationExportSettingsComponent implements OnInit {
...
           this.exportSettingsForm.controls.cccExportDate.clearValidators();
           this.exportSettingsForm.controls.cccExpenseState.clearValidators();
-          this.exportSettingsForm.controls.cccAccountName.clearValidators();
           this.exportSettingsForm.controls.cccExportType.setValue(null);
           this.exportSettingsForm.controls.cccExpenseState.setValue(null);
-          this.exportSettingsForm.controls.cccAccountName.setValue(null);
           this.exportSettingsForm.controls.cccExportGroup.setValue(null);
           this.exportSettingsForm.controls.cccExportDate.setValue(null);
         }
-        this.cccExportWatcher();
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts]: same as above

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 22]
```diff
@@ -379,17 +362,6 @@ export class ConfigurationExportSettingsComponent implements OnInit {
       this.cccWatcher();
     }
 
-    private setupDynamicOptions(): void {
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts]: why is this removed?

File: src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts [hunk 23]
```diff
@@ -467,50 +449,62 @@ export class ConfigurationExportSettingsComponent implements OnInit {
...
+        creditCardVendor: [(this.sageIntacctDefaultVendor?.find(creditCardVendor => creditCardVendor.id.toString() === this.exportSettings?.general_mappings.default_ccc_vendor.id)) || null],
+        creditCard: [(this.sageIntacctDefaultCreditCard?.find(creditCard => creditCard.id.toString() === this.exportSettings?.general_mappings.default_credit_card.id)) || null],
+        chargeCard: [(this.sageIntacctDefaultChargeCard?.find(chargeCard => chargeCard.id.toString() === this.exportSettings?.general_mappings.default_charge_card.id)) || null],
+        cccEntityName: [!(!!this.exportSettings?.configurations.reimbursable_expenses_object) ? this.exportSettings?.configurations.employee_field_mapping : null]
       });
+      if(this.exportSettings?.configurations.reimbursable_expenses_object==='BILL' || this.exportSettings?.configurations.reimbursable_expenses_object==='EXPENSE_REPORT') {
+        this.exportSettingsForm.get('employeeFieldMapping')?.disable();
+      }
+      if (this.exportSettings?.configurations.corporate_credit_card_expenses_object === 'CHARGE_CARD_TRANSACTION') {
```
Comment (ashwin1111) [inline src/app/shared/components/si/configuration/configuration-export-settings/configuration-export-settings.component.ts]: use enum for comparison everywhere
